### PR TITLE
implement a 'look' command like cpanm

### PR DESCRIPTION
**Pull request by tartansandal**

### CODE REVIEW

File: lib/Pinto/Shell.pm [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+);
+
+#------------------------------------------------------------------------------
+
+sub spawn {
+    my ($self) = @_;
+
+    my $shell = $self->shell;
+
+    my $cwd_guard = cwd_guard( $self->work_dir );
```
Comment (thaljef) [inline lib/Pinto/Shell.pm]: That worked out quite nicely, don't you think?

Comment (tartansandal) [inline lib/Pinto/Shell.pm]: Lovely :-)


Kahlil (Kal) Hodgson                       GPG: C9A02289
Head of Technology                         (m) +61 (0) 4 2573 0382
DealMax Pty Ltd

Suite 1416
401 Docklands Drive
Docklands VIC 3008 Australia

"All parts should go together without forcing.  You must remember that
the parts you are reassembling were disassembled by you.  Therefore,
if you can't get them together again, there must be a reason.  By all
means, do not use a hammer."  -- IBM maintenance manual, 1925

File: lib/App/Pinto/Command/look.pm [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  pinto --root=REPOSITORY_ROOT look [OPTIONS] TARGET ...
+
+=head1 DESCRIPTION
+
+Unpack one or more distributions and open directories with your shell.
+Handy if you want to manually review a distribution before use.
+
+=head1 COMMAND ARGUMENTS
+
+Arguments are the targets you wish to unpin.  Targets can be
```
Comment (thaljef) [inline lib/App/Pinto/Command/look.pm]: "unpin" -> "look at"

File: lib/Pinto/Action/Look.pm [hunk 3]
```diff
@@ -0,0 +1,75 @@
+# ABSTRACT: Force a package to stay in a stack
```
Comment (thaljef) [inline lib/Pinto/Action/Look.pm]: copy-n-pasted

File: lib/App/Pinto/Command/look.pm (hunk 2 above)
Comment (tartansandal) [inline lib/App/Pinto/Command/look.pm]: he he. nice catch.


Kahlil (Kal) Hodgson                       GPG: C9A02289
Head of Technology                         (m) +61 (0) 4 2573 0382
DealMax Pty Ltd

Suite 1416
401 Docklands Drive
Docklands VIC 3008 Australia

"All parts should go together without forcing.  You must remember that
the parts you are reassembling were disassembled by you.  Therefore,
if you can't get them together again, there must be a reason.  By all
means, do not use a hammer."  -- IBM maintenance manual, 1925

File: lib/Pinto/Action/Look.pm [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    isa      => TargetList,
+    traits   => [qw(Array)],
+    handles  => { targets => 'elements' },
+    required => 1,
+    coerce   => 1,
+);
+
+has shell => (
+    is       => 'ro',
+    isa      => Str,
```
Comment (thaljef) [inline lib/Pinto/Action/Look.pm]: Should probably use the `File` type from `Pinto::Types` here. Be sure to set `coerce => 1`

File: lib/Pinto/Shell.pm [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+package Pinto::Shell;
+
+use Moose;
+use MooseX::StrictConstructor;
+use MooseX::MarkAsMethods ( autoclean => 1 );
+use MooseX::Types::Moose qw(Str);
+
+use Pinto::Types qw(File Dir);
+use Pinto::Util qw(debug throw);
```
Comment (thaljef) [inline lib/Pinto/Shell.pm]: I don't see `debug` or `throw` being used anywhere.

**Comment by thaljef** [comment]
This looks really good, but we can't release it until it works reasonably with a remote repository. There are two choices for that:

1. Implement support for remote repositories by creating a subclass of `Pinto::Remote::Action` that fetches the remote tarball first (see `Pinto::Remote::Action::Install`l for example).

1. Make `App::Pinto::Command::look::execute()` bail out with an appropriate error message if the repository is remote (i.e. starts with /https?/). Then we'll add support for remote repositories later.

Either is fine with me.

**Comment by tartansandal** [comment]
Okay. Implemented 3 patches to fix copy-n-paste errors, 1 to make SHELL a
File <doh>, and 1 to die if the remote repository is remote (inspired by
the 'init' command). Opening distros from remote repositories makes good
sense, but I'm a little short on time ATM.

I'm using next::method() from mro 'cause its been so long and I can't
remember how to use SUPER anymore. Really wanted to use a 'before'
modifier, but App::Cmd does not seem to role that way. Added MRO::compat so
users of RHEL5 will be happier. Opinions most welcome ;-)


Kahlil (Kal) Hodgson                       GPG: C9A02289
Head of Technology                         (m) +61 (0) 4 2573 0382
DealMax Pty Ltd

Suite 1416
401 Docklands Drive
Docklands VIC 3008 Australia

"All parts should go together without forcing.  You must remember that
the parts you are reassembling were disassembled by you.  Therefore,
if you can't get them together again, there must be a reason.  By all
means, do not use a hammer."  -- IBM maintenance manual, 1925

**Comment by thaljef** [comment]
App::Cmd isn't based on Moo/Moose, so method modifiers are not an option there. I don't have anything against MRO (DBIx::Class is already using it). But for the sake of consistency, we have to choose one way or the other.  Currently we have no `next::method()` calls in the code but two of these:

```
App/Pinto/Command/help.pm:24:    my $rv = $self->SUPER::execute( $opts, $args );
App/Pinto/Command/nop.pm:29:    $self->SUPER::validate_args( $opts, $args );
```

**Comment by tartansandal** [comment]
I reckon we stick with the status quo -- now that I've been reminded of how
to use SUPER once again :-)

We don't have complex/multiple inheritance under App::Cmd so the whole MRO
thing is not important.

Will get onto a patch for this an hour or so.


Kahlil (Kal) Hodgson                       GPG: C9A02289
Head of Technology                         (m) +61 (0) 4 2573 0382
DealMax Pty Ltd

Suite 1416
401 Docklands Drive
Docklands VIC 3008 Australia

"All parts should go together without forcing.  You must remember that
the parts you are reassembling were disassembled by you.  Therefore,
if you can't get them together again, there must be a reason.  By all
means, do not use a hammer."  -- IBM maintenance manual, 1925

**Comment by thaljef** [comment]
> Will get onto a patch for this an hour or so.

Already done: bb113a

**Comment by thaljef** [comment]
I already merged this, but I just realized this command needs at `stack` option too.  Whenever a command accepts package names like Foo::Bar as a target, then it needs to know which stack to resolve that against, because every stack could have a different version of Foo::Bar.  This is an important feature of Pinto's behavior. I'll try to fix this so you can see what I mean.

**Comment by tartansandal** [comment]
Ah yes, that makes sense. Probably true of the sign command too.

**Comment by tartansandal** [comment]
Actually, ignore my last comment. When we sign, we sign all the authors
distributions. Hence the clone command to get a minimal set to sign.

**Comment by thaljef** [comment]
> When we sign, we sign all the authors distributions.

Different versions of a package could be owned by different authors. So I think the `sign` command still needs to resolve a package target to a specific distribution.

**Comment by thaljef** [comment]
Never mind that last comment. I see now that it signs *every* distribution by *every* author. So I'm going to delete the last part of this thread.

**Comment by thaljef** [comment]
I rewrote some of this in 09c626bd. Mostly I just re-arranged things. But I did make one interface change by removing the `--shell` command line option in favor of a `PINTO_SHELL` environment variable, which is consistent with other parts of the interface, such as `PINTO_EDITOR` and `PINTO_PAGER`. Thanks to you, this is a very nice addition to Pinto :smiley_cat:

**Comment by tartansandal** [comment]
Nice reorg. "Look"ing good ;-)


Kahlil (Kal) Hodgson                       GPG: C9A02289
Head of Technology                         (m) +61 (0) 4 2573 0382
DealMax Pty Ltd

Suite 1416
401 Docklands Drive
Docklands VIC 3008 Australia

"All parts should go together without forcing.  You must remember that
the parts you are reassembling were disassembled by you.  Therefore,
if you can't get them together again, there must be a reason.  By all
means, do not use a hammer."  -- IBM maintenance manual, 1925

**Pull request merged**